miden-lib/asm/miden/kernels/tx/tx.masm

```diff
@@ -251,14 +254,50 @@ proc.add_fungible_asset_to_note
     # => [note_ptr, note_idx]
 end
 
+#! Builds the stack into the NOTE_METADATA word, encoding the note type and execution hint
+#! into a single element.
+#!
+#! Inputs:  [tag, aux, note_type, execution_hint]
+#! Outputs: [NOTE_METADATA]
+proc.build_note_metadata
+    # validate the note type
+    # NOTE: encrypted notes are currently unsupported
+    dup.2 push.PRIVATE_NOTE eq dup.3 push.PUBLIC_NOTE eq or assert.err=ERR_INVALID_NOTE_TYPE
+    # => [tag, aux, note_type, execution_hint]
+
+    # copy data to validate the tag
+    dup.2 push.PUBLIC_NOTE dup.1 dup.3
+    # => [tag, note_type, public_note, note_type, tag, aux, note_type, execution_hint]
+
+    u32assert.err=ERR_NOTE_TAG_MUST_BE_U32
+    # => [tag, note_type, public_note, note_type, tag, aux, note_type, execution_hint]
+
+    # enforce the note type depending on the tag' bits
+    u32shr.30 push.ALL_NOTE_TYPES_ALLOWED eq cdrop assert_eq.err=ERR_NOTE_INVALID_TYPE_FOR_TAG
+    # => [tag, aux, note_type, execution_hint]
+
+    # populate the metadata
+    # swap movup.4 movup.4 
+    movup.3 movup.3 push.TWO_POW_38 mul add movup.2
+    # => [aux, encoded_type_and_ex_hint, tag]
+
+    # encode note_type and execution_hint into a single element
+    exec.account::get_id
+    # => [sender_acct_id, aux, encoded_type_and_ex_hint, tag]
```

Comment:
A couple comments here:

- Line 280 should probably be deleted.
- Comment on line 284 should probably be moved to line 279.
- A new comment should probably be written on line 284.